### ISSUE TITLE
Adding S1 NVDI operator

### ISSUE THREAD

**Issue by j08lue** [DHI-GRAS/processing_gpf#43]
From @ewamiti:

I have created this file for adding NDVI processor in QGIS from SNAP. 

```
NdviOp
s1tbx_Ndvi-Op
Ndvi operator
Radar
ParameterRaster|sourceProduct|The Input Bands|False|False
ParameterSelection|Red Factor|Red Factor|1.0
ParameterSelection|NIR factor|NIR Factor|1.0
ParameterBands|Red Factor|Red Band||sourceProduct|True
ParameterBands|NIR Factor|NIR Band||sourceProduct|True
OutputRaster|-out|Output Image
```

Unfortunately when i run i get this error :

```
2017-05-28T11:44:11 1 Traceback (most recent call last):
File "C:/Users/Research_2/.qgis2/python/plugins\processing\gui\AlgorithmDialog.py", line 228, in accept
self.setParamValues()
File "C:/Users/Research_2/.qgis2/python/plugins\processing\gui\AlgorithmDialog.py", line 107, in setParamValues
param, self.mainWidget.valueItems[param.name]):
File "C:/Users/Research_2/.qgis2/python/plugins\processing\gui\AlgorithmDialog.py", line 150, in setParamValue
return param.setValue(widget.currentIndex())
AttributeError: 'GPFBandsSelectorPanel' object has no attribute 'currentIndex'
```

**Comment by j08lue** [comment]
This error has to do with our integration with the QGIS processing plugin. 

@ewamiti Please answer the following:
1. Which version of QGIS are you using?
1. Are you using a custom version of the processing toolbox (check if `~\.qgis2\python\plugins\processing` exists), e.g. installed with WOIS? If so, which one (check the `metadata.txt`)?

Our [`GPFBandsSelectorPanel`](https://github.com/DHI-RAS/processing_gpf/blob/9fd2b1743e026c4c4a242415e373ba111e399fc4/GPFParametersPanel.py#L126-L162) is lacking the `currentIndex` method that is needed [in QGIS' GUI dialogue](https://github.com/qgis/QGIS/blob/final-2_18_9/python/plugins/processing/gui/AlgorithmDialog.py#L149-L150).

@radosuav, could you assist in figuring out how we fix this? What does this `currentIndex` method return? There is one example [here](https://github.com/qgis/QGIS/blob/c77172ed3d50800c80c9165daa061dbe5b6a63c3/python/plugins/db_manager/dlg_table_properties.py#L306-L313), but I am not sure how that can be transferred to our case.

**Comment by ewamiti** [comment]
Thank you
I am using QGIS 2.18.3.
I find the text below in the meta data.
[general]
name=Processing
description=Spatial data processing framework for QGIS
about=Spatial data processing framework for QGIS
category=Analysis
version=2.12.99
qgisMinimumVersion=2.13

author=Victor Olaya
email=volayaf@gmail.com

icon=images/alg.png

homepage=http://qgis.org
tracker=https://hub.qgis.org/projects/quantum-gis/issues
repository=https://github.com/qgis/QGIS


Thank you

**Comment by radosuav** [comment]
@ewamiti good try :) However, there are some problems in your description file, including one which is causing the exception. The correct operator should look like this:
```
NdviOp
SNAP_NdviOp
NDVI operator
Optical
ParameterRaster|sourceProduct|Input image|False|False
ParameterNumber|redFactor|Red Factor|0.0|1000.0|1.0
ParameterNumber|nirFactor|NIR Factor|0.0|1000.0|1.0
ParameterBands|redSourceBand|Red  source band||sourceProduct|False
ParameterBands|nirSourceBand|NIR source band||sourceProduct|False
OutputRaster|-out|Output Image
```
Going from the top line by line:
1. OK
2. S1 is not an optical satellite so S1TBX does not have NDVI operator. If you open the NDVI operator in SNAP and go to Help > About then under "Full name" it is listed as belonging to org.esa.snap. This means that it's a generic SNAP operator (not belonging to any toolbox). Therefore the name should be SNAP_NdviOp. Also the description file should be saved in processing_gpf\snap_generic_description folder.
![screenshot](https://cloud.githubusercontent.com/assets/4759850/26540754/2798822e-4453-11e7-986d-890dcaca9580.png)
3. Just capitalized NDVI but otherwise OK.
4. In SNAP this algorithm can be found under "Optical" in the main menu.
5. I renamed the parameter for clarity but this is subjective (could also be called Source Product, etc.). Otherwise good.
6. Firstly, the internal parameter name (second item in this line) should exactly correspond to the parameter name within the NDVI operator, i.e. it should be redFactor.
![screenshot](https://cloud.githubusercontent.com/assets/4759850/26541177/f5712088-4454-11e7-98b4-3b83ab2052fa.png)
Secondly, in the SNAP GUI this is not a selection widget but rather a number input widget. So I changed it to ParameterNumber with a range from 0 to 1000 and default value of 1.0.
![screenshot](https://cloud.githubusercontent.com/assets/4759850/26541203/1204c092-4455-11e7-9dbc-bfd5f0178754.png)
7. Same as line 6.
8. The exception was occurring because this parameter had the same internal name as the parameter on line 6 (Red Factor) which was originally a ParameterSelection. Therefore it was created with a widget of ParameterBands but internal workings of ParameterSelection and this led to the exception. After giving it the correct internal name (corresponding to the name within the NDVI operator) the exception is gone. I also changed it's external name to correspond to the name shown in SNAP (Red source band) and changed it from optional to compulsory (changing last True to False).
9. Same as line 8.
10. OK

Hope all is clear but otherwise let me know.

**Comment by ewamiti** [comment]
Thank you.
have tried it and it works. My parameters display window for NDVI had only infromation about factors.

**Comment by j08lue** [comment]
> have tried it and it works

@radosuav should we add this algorithm to the distribution then?

**Comment by ewamiti** [comment]
Oh yes. That would be great!

**Comment by ewamiti** [comment]
I have others that i created i will try running then and give you a feedback.

**Comment by radosuav** [comment]
@j08lue good idea. Should it be a pull request from @ewamiti?

@ewamiti sometimes not all the parameters are displayed until you open a file in the I/O Parameters tab.

**Comment by j08lue** [comment]
@ewamiti If you could put together a pull request, that would be great. In case you are not familiar with the procedure - you need to [fork this repository](https://github.com/DHI-GRAS/processing_gpf#fork-destination-box), commit and push to your fork, and then return to this repo on GitHub where it will show the option to open a PR.

**Comment by ewamiti** [comment]
I am not really getting the steps right.

**Comment by j08lue** [comment]
I can see you made the fork at https://github.com/ewamiti/processing_gpf and committed to it. You should now on the GitHub page for your fork see the option to make a pull request, don't you?

![image](https://cloud.githubusercontent.com/assets/3404817/26570514/496eeff0-4511-11e7-82c5-7034cb56ca73.png)

Click on that and write a few comments and your PR should be ready for us to review and merge.

**Comment by j08lue** [comment]
NB: Even after you opened the PR, you can still push to your fork and the commits will be appended to the PR.

**Comment by ewamiti** [comment]
I have put through a request

**Comment by ewamiti** [comment]
Was this algorithm added to the distribution? I can not trace it.

**Comment by j08lue** [comment]
Hi @ewamiti sorry for not responding earlier. Please see @radosuav's [comment on your PR](https://github.com/DHI-GRAS/processing_gpf/pull/45#issuecomment-311088763) and let us know (on that thread) if you can make the changes he asked for.

**Comment by radosuav** [comment]
@ewamiti, your NDVI implementation has now been merged (see https://github.com/DHI-GRAS/processing_gpf/pull/47) and will be included in version 1.4.3 of the plugin, which should be released very soon.
